### PR TITLE
Update ContrabutionPage schema/bao to use best-practices

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
BAO/metadata updates for the ContributionPage entity

Technical Details
----------------------------------------
- Not null date fields
- Callbacks for default values
- current_timestamp() field defaults
- Deprecate BAO::create function

### CODE REVIEW

File: CRM/Contribute/BAO/ContributionPage.php [hunk 1]
```diff
@@ -26,17 +26,12 @@
...
-   *
+   * @deprecated
    * @param array $params
-   *
    * @return CRM_Contribute_DAO_ContributionPage
    */
   public static function create($params) {
-    // @todo  - this implode is probably handled by writeRecord - test & remove.
-    if (isset($params['payment_processor']) && is_array($params['payment_processor'])) {
-      $params['payment_processor'] = implode(CRM_Core_DAO::VALUE_SEPARATOR, $params['payment_processor']);
```
Comment (colemanw) [inline CRM/Contribute/BAO/ContributionPage.php:38]: Indeed it is handled by writeRecord. And there is a test:
https://github.com/civicrm/civicrm-core/blob/4772bbd375c116226cd23a38a423fdf8c576fc71/tests/phpunit/api/v3/ContributionPageTest.php#L224-L226

File: tests/phpunit/CRM/Contribute/BAO/ContributionPageTest.php [hunk 2]
```diff
@@ -15,35 +15,6 @@
...
-      'start_date' => '20091022105900',
-      'start_date_time' => '10:59AM',
-      'end_date' => '19700101000000',
-      'end_date_time' => '',
-      'is_credit_card_only' => '',
-    ];
-
-    $contributionPageID = CRM_Contribute_BAO_ContributionPage::create($params)->id;
-    $this->assertIsInt($contributionPageID);
-    $this->callAPISuccess('ContributionPage', 'delete', ['id' => $contributionPageID]);
```
Comment (colemanw) [inline tests/phpunit/CRM/Contribute/BAO/ContributionPageTest.php:44]: It's hard to think of a more useless test

Comment (demeritcowboy) [reply tests/phpunit/CRM/Contribute/BAO/ContributionPageTest.php:44]: I agree that just checking that id is an integer is silly, but the test would still have been useful as a sanity check to expose some typo or undefined var that gets introduced, or other error. But if writeRecord is tested with ContributionPage then there's no need for this now that create is deprecated.

Comment (demeritcowboy) [reply tests/phpunit/CRM/Contribute/BAO/ContributionPageTest.php:44]: Sometimes different params have different code paths or trigger different errors. But overall I'm agreeing with you.

**Pull request merged**